Automattic/sensei
# Create generic enrolment job scheduler and introduce course recalculation

This is a larger PR that accomplishes a few things. I started working on the task to introduce an async job to recalculate a course's enrolment and I realized there was going to be a lot shared with @gkaragia's learner calculation async job. I had a chat with him and it seemed appropriate to split his class in two, one for the job itself and one for the scheduling. For the scheduling class, I've made it responsible for handling all enrolment async job handling. We could make this Sensei-wide, but for now I've opted to limit the scope of responsibility for the class. A few other smaller additions were also added to accomplish these things.

Just as a note: nothing currently triggers this async job. I've added a snippet to trigger manually. No user-facing changes should have been made in this PR.

### Changes Proposed
- Split `Sensei_Enrolment_Calculation_Scheduler` class from #2874 and introduce a shared scheduler class. (This also involved some test updates)
- Introduce a concept of a course enrolment hash. Similar to the site wide hash, this is used to easily invalidate all enrolment results for a course. In order to quickly invalidate results for all users, I use `$wpdb->update()`. This will work in tandem with that to prevent cached results from being considered.
- Adds a helper to get all the user IDs for enrolled students in a course (\Sensei_Course_Enrolment::get_enrolled_user_ids).
- Finally, it adds a job to recalculate the enrolment for a single course. This can be done for either all users _or_ just already enrolled students.
  - An example of when you might want to just run it for enrolled students is when you remove a product from a course but don't add any new ones.

### Testing Instructions
You can use this snippet. 
```php
add_action( 'transition_post_status', function( $to, $from, $post ) { 
	if ( $from === $to || 'course' !== $post->post_type ) {
		return;
	}
	
	$course_enrolment = Sensei_Course_Enrolment::get_course_instance( $post->ID );
	$course_enrolment->recalculate_enrolment( false ); // also try true
}, 10, 3 );
```

Unpublish a course and then publish it again. When you unpublish, you should notice all the `sensei_course_enrolment_{COURSEID}` user meta get cleared. They won't recalculate because draft courses short circuit early for not enrolled (however, the term assignments will go away after the job runs). Publish again and the job will trigger and they'll get calculated again.

We should also test to make sure there wasn't a regression from #2874.
